### PR TITLE
[SYCL] Pass buffer_location property to buffer

### PR DESCRIPTION
**Pull request by maximdimakov**
There is the accessor property "buffer_location" that allows to allocate buffer in definite location (spec: https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/supported/sycl_ext_intel_buffer_location.asciidoc)
Current implementation doesn't implement allocating a buffer at the passed location and the buffer will be re-sided when a kernel is enqueued. It leads to problems of various kinds.
The proposed solution implies adding new buffer property in order to store it in a buffer and use it when the buffer is allocated.
If the property is not supported by device it will be ignored.
Signed-off-by: mdimakov [maxim.dimakov@intel.com](mailto:maxim.dimakov@intel.com)

### CODE REVIEW

File: sycl/test/abi/vtable.cpp
Comment (maximdimakov) [reply sycl/test/abi/vtable.cpp]: Remove virtual functions

**Comment by maximdimakov** [comment]
The test `fpga_tests/buffer_location.cpp` failed because the device the test ran on has no support for the buffer_location feature. There 2 ways for resolve this problem:
- Disable this test until CI devices get support for the buffer_location feature
- Add sycl::aspect for this feature and check it. The difficulty is that spec has no mention for such aspect. It can be added there, but this action may require indefinite amount of time

**Comment by bader** [comment]
> * Add sycl::aspect for this feature and check it. The difficulty is that spec has no mention for such aspect. It can be added there, but this action may require indefinite amount of time

I suggest filing a bug report on the spec and ask for fix ETA.

Please, update the link to the spec and fix punctuation in the description.

**Comment by bader** [comment]
> > * Add sycl::aspect for this feature and check it. The difficulty is that spec has no mention for such aspect. It can be added there, but this action may require indefinite amount of time
> 
> I suggest filing a bug report on the spec and ask for fix ETA.
> 
> Please, update the link to the spec and fix punctuation in the description.

Just confirming that this the issue in the spec and it's going to be fixed by adding a new device aspect might be enough to implement a proper test fix.

File: sycl/source/detail/context_impl.hpp
Comment (s-kanaev) [inline sycl/source/detail/context_impl.hpp:183]: It's better to have enum or a bool with human-readable values rather than a mere integer. Otherwise, these are magic numbers which aren't intuitive to understand.

File: sycl/source/detail/context_impl.hpp
Comment (s-kanaev) [inline sycl/source/detail/context_impl.hpp:184]:
```suggestion
  mutable int MSupportBufferLocationByDevices;
```

File: sycl/source/detail/context_impl.hpp
Comment (s-kanaev) [inline sycl/source/detail/context_impl.hpp:171]:
```suggestion
  bool isBufferLocationSupported() const; 
```

Can we have it's implementation in header so that compiler can inline a function at caller's side if possible?

File: sycl/source/detail/context_impl.cpp
Comment (s-kanaev) [inline sycl/source/detail/context_impl.cpp:215]:
```suggestion
  // Check that devices within context have support of buffer location
```

File: sycl/include/CL/sycl/detail/property_helper.hpp
Comment (s-kanaev) [inline sycl/include/CL/sycl/detail/property_helper.hpp:51]:
```suggestion
  PropWithDataKindSize = 6,
```

**Review by s-kanaev — CHANGES_REQUESTED** [review]

File: sycl/unittests/buffer/BufferLocation.cpp
Comment (maximdimakov) [reply sycl/unittests/buffer/BufferLocation.cpp:165]: There is no kernel in the command group above, so `piMemBufferCreate` will not be called. This check verifies that new buffer_location is passed to the buffer properties. Check that this value is passed correctly is placed above
https://github.com/intel/llvm/pull/5604/files/4d548846221b21799da61a62f7755bc33b59f829#diff-8e8ed34037520ee5405a46f25d39b96ff066f684a71560371fd3258ac46be9f8R146

File: sycl/source/detail/context_impl.hpp
Comment (maximdimakov) [reply sycl/source/detail/context_impl.hpp]: Replaced it by enum

File: sycl/source/detail/context_impl.hpp
Comment (maximdimakov) [reply sycl/source/detail/context_impl.hpp]: Replaced it by enum

File: sycl/source/detail/context_impl.hpp
Comment (maximdimakov) [reply sycl/source/detail/context_impl.hpp]: This function is not simple enough to be in header

**Comment by maximdimakov** [comment]
@bader I have opened PR with adding the aspect https://github.com/intel/llvm/pull/5660. It was decided to ignore this property if it is not supported on device. So, fpga_tests/buffer_location.cpp works as before.

File: sycl/source/detail/context_impl.cpp
Comment (sherry-yuan) [inline sycl/source/detail/context_impl.cpp:227]: I think this specific way of calling get device info will always return false since the lower level runtime does not support passing `PI_MEM_PROPERTIES_ALLOC_BUFFER_LOCATION` into the get device info API. Is there no other way of determining whether the target is an FPGA?

If we do this, then the OpenCL Spec has to be changed.

Comment (sherry-yuan) [reply sycl/source/detail/context_impl.cpp:227]: Although technically the opencl spec is written in a way that enforces all vendor's runtime to accept this property, so technically it should be fine to directly pass it into other runtime (although they may do nothing with the property).

Comment (maximdimakov) [reply sycl/source/detail/context_impl.cpp:227]: `Is there no other way of determining whether the target is an FPGA that does not rely on lower level runtime?`
We can determine if the device is FPGA or not in runtime. I tried to do this in https://github.com/intel/llvm/pull/5604/commits/ccf73a63822962464eec97433e86daf8d4cbacbb
Do you think it would be better to just check if a device is FPGA or not?

Comment (sherry-yuan) [reply sycl/source/detail/context_impl.cpp:227]: Yes I do think it is better to check if a device is FPGA directly, because even if it error out FPGA device, it will prompt the other vendor to support passing in of buffer location feature in their runtime API.

File: sycl/include/CL/sycl/detail/property_helper.hpp
Comment (sherry-yuan) [inline sycl/include/CL/sycl/detail/property_helper.hpp:51]: Is there any reason why new enum is not added at the end?

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: Feel free to ignore, I found the reason: there is a comparison somewhere where all other enum of the same type cannot exceed this last enum.

https://github.com/intel/llvm/blob/54ede407edeb93b7e9334d1e725f48bf981b2965/sycl/include/CL/sycl/detail/property_list_base.hpp#L94

Comment (sherry-yuan) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: Is this an ABI breaking change given that PropWithDataKindSize  changed from 5 -> 6?

Comment (maximdimakov) [reply sycl/include/CL/sycl/detail/property_helper.hpp:51]: It is not an ABI breaking change, because it is not a property and it is used only in `property_list_base.hpp`

**Comment by maximdimakov** [comment]
@s-kanaev could you, please, review?

**Review by sherry-yuan — APPROVED** [review]
LGTM, The latest change passes internal test and has the expected behavior. 
Wait for others' approval though.

File: sycl/source/detail/context_impl.hpp
Comment (s-kanaev) [inline sycl/source/detail/context_impl.hpp:185]:
```suggestion
  mutable PropertySupport MSupportBufferLocationByDevices;
```

**Review by s-kanaev — CHANGES_REQUESTED** [review]
LGTM overall

File: sycl/unittests/CMakeLists.txt
Comment (s-kanaev) [inline sycl/unittests/CMakeLists.txt:36]: That's an interesting change in this late commit...

**Review by s-kanaev — APPROVED** [review]
LGTM

File: sycl/unittests/CMakeLists.txt
Comment (maximdimakov) [reply sycl/unittests/CMakeLists.txt:36]: I tried to resolve merge conflict. This line has already added to CMakeLists.txt

**Pull request merged**
